### PR TITLE
Add support for ternary match kind to eBPF backend

### PR DESCRIPTION
**Pull request by osinstom**
Co-authored-by: Mateusz Kossakowski <mateusz.kossakowski@orange.com>
Co-authored-by: Jan Palimąka <jan.palimaka@orange.com>

### CODE REVIEW

File: backends/ebpf/ebpfOptions.cpp [hunk 1]
```diff
@@ -26,6 +26,14 @@ EbpfOptions::EbpfOptions() {
         registerOption("--trace", nullptr,
                 [this](const char*) { emitTraceMessages = true; return true; },
                 "Generate tracing messages of packet processing");
+        registerOption("--max-ternary-masks", "MAX_TERNARY_MASKS",
+                [this](const char *arg) {
+                   unsigned int parsed_val = std::strtoul(arg, nullptr, 0);
+                   if (parsed_val >= 2)
+                       this->maxTernaryMasks = parsed_val;
```
Comment (tatry) [inline backends/ebpf/ebpfOptions.cpp:33]: Maybe some error when option is ignored?

File: backends/ebpf/psa/README.md [hunk 2]
```diff
@@ -414,6 +511,11 @@ with some NICs. So far, we have verified the correct behavior with Intel 82599ES
 - DirectCounter and DirectMeter externs are not supported for P4 tables with implementation (ActionProfile).
 - The `xdp2tc=head` mode works only for packets larger than 34 bytes (the size of Ethernet and IPv4 header).
 - `value_set` only supports the exact match type and can only match on a single field in the `select()` expression.
+- The number of entries in ternary tables are limited by the number of unique ternary masks. If a P4 program uses many ternary tables and the `--max-ternary-masks` (default: 128) is set 
+  to a high value, the P4 program may not compile due to the BPF complexity issue (the 1M instruction limit exceeded). This is the limitation of the current implementation of the TSS algorithm that
```
Comment (tatry) [inline backends/ebpf/psa/README.md]:
```suggestion
  to a high value, the P4 program may not load into BPF subsystem due to the BPF complexity issue (the 1M instruction limit exceeded). This is the limitation of the current implementation of the TSS algorithm that
```

File: backends/ebpf/psa/README.md [hunk 3]
```diff
@@ -414,6 +511,11 @@ with some NICs. So far, we have verified the correct behavior with Intel 82599ES
...
+  requires iteration over BPF maps. Note that the recent kernel introduced the [bpf_for_each_map_elem()](https://lwn.net/Articles/846504/) helper that should simplify the iteration process and help to overcome the current limitation.
+- Setting a size of ternary tables does not currently work.
```
Comment (tatry) [inline backends/ebpf/psa/README.md:519]: In some way it works for tuples, not whole table

File: backends/ebpf/tests/p4testdata/psa-ternary.p4 [hunk 4]
```diff
@@ -0,0 +1,192 @@
...
+
+control EgressDeparserImpl(packet_out buffer,
+                           out empty_t clone_e2e_meta,
+                           out empty_t recirculate_meta,
+                           inout headers hdr,
+                           in metadata meta,
+                           in psa_egress_output_metadata_t istd,
+                           in psa_egress_deparser_input_metadata_t edstd)
+{
+    CommonDeparserImpl() cp;
```
Comment (kmateuszssak) [inline backends/ebpf/tests/p4testdata/psa-ternary.p4:195]: Could we don't use `CommonDeparserImpl`? It will be few lines code fewer.

**Review by kmateuszssak — APPROVED** [review]
Ok.

File: backends/ebpf/ebpfTable.h [hunk 5]
```diff
@@ -72,7 +72,11 @@ class EBPFTable : public EBPFTableBase {
  protected:
     const cstring prefixFieldName = "prefixlen";
 
-    bool isLPMTable();
+    bool isLPMTable() const;
+    bool isTernaryTable() const;
+
+    void emitTernaryInstance(CodeBuilder* builder);
```
Comment (tatry) [inline backends/ebpf/ebpfTable.h:78]: Maybe `emitTableInstance`? It is not clear why this name.

File: backends/ebpf/psa/ebpfPsaTable.cpp [hunk 6]
```diff
@@ -293,51 +293,35 @@ ActionTranslationVisitor* EBPFTablePSA::createActionTranslationVisitor(
...
 }
 
 void EBPFTablePSA::emitInstance(CodeBuilder *builder) {
-    if (keyGenerator != nullptr) {
+    if (isTernaryTable()) {
+        emitTernaryInstance(builder);
+    } else {
         TableKind kind = isLPMTable() ? TableLPMTrie : TableHash;
-        emitTableDecl(builder, instanceName, kind,
+        builder->target->emitTableDecl(builder, instanceName, kind,
```
Comment (tatry) [inline backends/ebpf/psa/ebpfPsaTable.cpp:312]: What about spinlock for meters?

File: backends/ebpf/ebpfTable.cpp [hunk 7]
```diff
@@ -105,23 +105,47 @@ void EBPFTable::initKey() {
 }
 
 // Performs the following validations:
-// 1. Validates if LPM key is the last one from match keys (ignores selector fields).
+// - Validates if LPM key is the last one from match keys in an LPM table (ignores selector fields).
+// - Validates if match fields in ternary tables are sorted by size
+//   in descending order (ignores selector fields).
```
Comment (tatry) [inline backends/ebpf/ebpfTable.cpp:110]: It might be worth to mention this in documentation

File: backends/ebpf/ebpfTable.cpp [hunk 8]
```diff
@@ -693,6 +904,21 @@ bool EBPFTable::isLPMTable() {
     return isLPM;
 }
 
+bool EBPFTable::isTernaryTable() const {
+    if (keyGenerator != nullptr) {
+        // If any key field is LPM we will generate an LPM table
```
Comment (tatry) [inline backends/ebpf/ebpfTable.cpp]:
```suggestion
        // If any key field is ternary we will generate a ternary table
```

**Review by tatry — COMMENTED** [review]
Tests for ternary table with ActionSelector and ActionProfile are missing.

File: backends/ebpf/psa/ebpfPsaTable.cpp (hunk 6 above)
Comment (osinstom) [reply backends/ebpf/psa/ebpfPsaTable.cpp:312]: Can you clarify?

File: backends/ebpf/ebpfTable.h (hunk 5 above)
Comment (osinstom) [reply backends/ebpf/ebpfTable.h:78]: keep in mind that this is a method of "table" object. So, we shouldn't have a call to table.emitTableInstance()

File: backends/ebpf/psa/ebpfPsaTable.cpp (hunk 6 above)
Comment (tatry) [reply backends/ebpf/psa/ebpfPsaTable.cpp:312]: I see that tests are passing, but I don't find a place where spinlocks are emitted for DirectMeters.

**Pull request closed without merging**